File: bench/polyglot/python/time_flow.py
#!/usr/bin/env python3
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone


NANOS_PER_SECOND = 1_000_000_000


@dataclass(frozen=True)
class Period:
    months: int = 0
    days: int = 0
    nanoseconds: int = 0

    def __add__(self, other: Period | Instant) -> Period | Instant:
        if isinstance(other, Instant):
            return other + self
        return Period(
            self.months + other.months,
            self.days + other.days,
            self.nanoseconds + other.nanoseconds,
        )

    def total_nanoseconds(self) -> int:
        if self.months != 0:
            raise ValueError("calendar period has no fixed total")
        return self.days * 86_400 * NANOS_PER_SECOND + self.nanoseconds


@dataclass(frozen=True)
class Instant:
    seconds: int
    nanosecond: int = 0

    @staticmethod
    def utc(
        year: int,
        month: int,
        day: int,
        hour: int = 0,
        minute: int = 0,
        second: int = 0,
        nanosecond: int = 0,
    ) -> Instant:
        dt = datetime(year, month, day, hour, minute, second, tzinfo=timezone.utc)
        return Instant(int(dt.timestamp()), nanosecond)

    @staticmethod
    def from_unix_ms(milliseconds: int) -> Instant:
        seconds = milliseconds // 1000
        nanos = (milliseconds - seconds * 1000) * 1_000_000
        return Instant(seconds, nanos)

    @staticmethod
    def from_unix_ns(nanoseconds: int) -> Instant:
        seconds = nanoseconds // NANOS_PER_SECOND
        nanos = nanoseconds - seconds * NANOS_PER_SECOND
        return Instant(seconds, nanos)

    @staticmethod
    def parse_iso8601(text: str) -> Instant:
        prefix, offset = text[:-6], text[-6:]
        sign = 1 if offset[0] == "+" else -1
        offset_seconds = sign * (int(offset[1:3]) * 3600 + int(offset[4:6]) * 60)
        head, frac = prefix.split(".")
        dt = datetime.strptime(head, "%Y-%m-%dT%H:%M:%S").replace(
            tzinfo=timezone.utc
        )
        frac = (frac + "000000000")[:9]
        return Instant(int(dt.timestamp()) - offset_seconds, int(frac))

    def fields(self) -> datetime:
        return datetime.fromtimestamp(self.seconds, tz=timezone.utc)

    def year(self) -> int:
        return self.fields().year

    def month(self) -> int:
        return self.fields().month

    def day(self) -> int:
        return self.fields().day

    def hour(self) -> int:
        return self.fields().hour

    def minute(self) -> int:
        return self.fields().minute

    def second(self) -> int:
        return self.fields().second

    def iso8601(self) -> str:
        dt = self.fields()
        text = dt.strftime("%Y-%m-%dT%H:%M:%S")
        if self.nanosecond != 0:
            frac = f"{self.nanosecond:09d}".rstrip("0")
            text += f".{frac}"
        return text + "Z"

    def unix_nanoseconds(self) -> int:
        return self.seconds * NANOS_PER_SECOND + self.nanosecond

    def __add__(self, period: Period) -> Instant:
        dt = self.fields()
        total_month = dt.year * 12 + (dt.month - 1) + period.months
        year, month0 = divmod(total_month, 12)
        month = month0 + 1
        day = min(dt.day, days_in_month(year, month))
        shifted = datetime(
            year,
            month,
            day,
            dt.hour,
            dt.minute,
            dt.second,
            tzinfo=timezone.utc,
        )
        shifted += timedelta(days=period.days)
        total_nanos = self.nanosecond + period.nanoseconds
        seconds_delta, nanos = divmod(total_nanos, NANOS_PER_SECOND)
        shifted += timedelta(seconds=seconds_delta)
        return Instant(int(shifted.timestamp()), nanos)

    def __sub__(self, other: Instant | Period) -> Period | Instant:
        if isinstance(other, Period):
            return self + Period(-other.months, -other.days, -other.nanoseconds)
        return Period(nanoseconds=self.unix_nanoseconds() - other.unix_nanoseconds())

    def __gt__(self, other: Instant) -> bool:
        return (self.seconds, self.nanosecond) > (other.seconds, other.nanosecond)


def days_in_month(year: int, month: int) -> int:
    if month == 2:
        leap = year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)
        return 29 if leap else 28
    return [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month - 1]


def months(value: int) -> Period:
    return Period(months=value)


def days(value: int) -> Period:
    return Period(days=value)


def seconds(value: int) -> Period:
    return Period(nanoseconds=value * NANOS_PER_SECOND)


def nanoseconds(value: int) -> Period:
    return Period(nanoseconds=value)


def main() -> int:
    checksum = 0
    anchor = Instant.parse_iso8601("2026-06-17T03:04:05.123456789+03:00")

    for i in range(12000):
        base = Instant.utc(
            2020 + (i % 7),
            1 + (i % 12),
            25 + (i % 4),
            hour=i % 24,
            minute=(i * 7) % 60,
            second=(i * 11) % 60,
            nanosecond=(i * 1_234_567) % 1_000_000_000,
        )
        period = months((i % 15) + 1)
        period = period + days(i % 21)
        period = period + seconds(i % 3600)
        period = period + nanoseconds((i % 1000) * 1000)
        shifted = base + period
        elapsed = shifted - base
        epoch = Instant.from_unix_ms(1_700_000_000_000 + (i * 37))
        tiny = Instant.from_unix_ns(1_000_000_000 + (i * 1_000_003) + 999)
        clamp = Instant.utc(2024, 1, 31) + months((i % 3) + 1)
        probe = seconds(5) + days(1) + anchor

        checksum += shifted.year() * 37
        checksum += shifted.month() * 31
        checksum += shifted.day() * 29
        checksum += shifted.hour() * 23
        checksum += shifted.minute() * 19
        checksum += shifted.second() * 17
        checksum += shifted.nanosecond % 1_000_003
        checksum += elapsed.total_nanoseconds() % 1_000_003
        checksum += epoch.second() + epoch.nanosecond % 1009
        checksum += tiny.unix_nanoseconds() % 9973
        checksum += clamp.day() * 13 + clamp.month()
        checksum += probe.day() + probe.second()
        if i % 97 == 0:
            checksum += len(shifted.iso8601())
        checksum += 7 if shifted > base else 3
        checksum %= 2_147_483_647

    return checksum


if __name__ == "__main__":
    print(main())

File: bench/polyglot/ruby/time_flow.rb
#!/usr/bin/env ruby

NANOS_PER_SECOND = 1_000_000_000

class Period
  attr_reader :months, :days, :nanoseconds

  def initialize(months = 0, days = 0, nanoseconds = 0)
    @months = months
    @days = days
    @nanoseconds = nanoseconds
  end

  def +(other)
    if other.is_a?(Instant)
      other + self
    else
      Period.new(
        @months + other.months,
        @days + other.days,
        @nanoseconds + other.nanoseconds
      )
    end
  end

  def total_nanoseconds
    raise "calendar period has no fixed total" unless @months == 0

    @days * 86_400 * NANOS_PER_SECOND + @nanoseconds
  end
end

class Instant
  include Comparable

  attr_reader :seconds, :nanosecond

  def initialize(seconds, nanosecond = 0)
    @seconds = seconds
    @nanosecond = nanosecond
  end

  def self.utc(year, month, day, hour = 0, minute = 0, second = 0,
               nanosecond = 0)
    Instant.new(Time.utc(year, month, day, hour, minute, second).to_i,
                nanosecond)
  end

  def self.from_unix_ms(milliseconds)
    seconds, millis = milliseconds.divmod(1000)
    Instant.new(seconds, millis * 1_000_000)
  end

  def self.from_unix_ns(nanoseconds)
    seconds, nanos = nanoseconds.divmod(NANOS_PER_SECOND)
    Instant.new(seconds, nanos)
  end

  def self.parse_iso8601(text)
    prefix = text[0...-6]
    offset = text[-6..]
    sign = offset[0] == '+' ? 1 : -1
    offset_seconds = sign * (offset[1, 2].to_i * 3600 + offset[4, 2].to_i * 60)
    head, frac = prefix.split('.')
    parts = head.scan(/\d+/).map(&:to_i)
    nanos = (frac + '000000000')[0, 9].to_i
    local = Time.utc(parts[0], parts[1], parts[2], parts[3], parts[4], parts[5])
    Instant.new(local.to_i - offset_seconds, nanos)
  end

  def fields
    Time.at(@seconds).utc
  end

  def year
    fields.year
  end

  def month
    fields.month
  end

  def day
    fields.day
  end

  def hour
    fields.hour
  end

  def minute
    fields.min
  end

  def second
    fields.sec
  end

  def iso8601
    text = fields.strftime('%Y-%m-%dT%H:%M:%S')
    if @nanosecond != 0
      frac = format('%09d', @nanosecond).sub(/0+$/, '')
      text += ".#{frac}"
    end
    text + 'Z'
  end

  def unix_nanoseconds
    @seconds * NANOS_PER_SECOND + @nanosecond
  end

  def +(period)
    base = fields
    total_month = base.year * 12 + (base.month - 1) + period.months
    year, month0 = total_month.divmod(12)
    month = month0 + 1
    day = [base.day, days_in_month(year, month)].min
    shifted = Time.utc(year, month, day, base.hour, base.min, base.sec)
    shifted += period.days * 86_400
    seconds_delta, nanos = (@nanosecond + period.nanoseconds).divmod(
      NANOS_PER_SECOND
    )
    shifted += seconds_delta
    Instant.new(shifted.to_i, nanos)
  end

  def -(other)
    if other.is_a?(Period)
      self + Period.new(-other.months, -other.days, -other.nanoseconds)
    else
      Period.new(0, 0, unix_nanoseconds - other.unix_nanoseconds)
    end
  end

  def <=>(other)
    [@seconds, @nanosecond] <=> [other.seconds, other.nanosecond]
  end
end

def days_in_month(year, month)
  if month == 2
    return ((year % 4).zero? && (!(year % 100).zero? || (year % 400).zero?)) ? 29 : 28
  end
  [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month - 1]
end

def months(value)
  Period.new(value, 0, 0)
end

def days(value)
  Period.new(0, value, 0)
end

def seconds(value)
  Period.new(0, 0, value * NANOS_PER_SECOND)
end

def nanoseconds(value)
  Period.new(0, 0, value)
end

def main
  checksum = 0
  anchor = Instant.parse_iso8601('2026-06-17T03:04:05.123456789+03:00')

  12_000.times do |i|
    base = Instant.utc(
      2020 + (i % 7),
      1 + (i % 12),
      25 + (i % 4),
      i % 24,
      (i * 7) % 60,
      (i * 11) % 60,
      (i * 1_234_567) % 1_000_000_000
    )
    period = months((i % 15) + 1)
    period += days(i % 21)
    period += seconds(i % 3600)
    period += nanoseconds((i % 1000) * 1000)
    shifted = base + period
    elapsed = shifted - base
    epoch = Instant.from_unix_ms(1_700_000_000_000 + (i * 37))
    tiny = Instant.from_unix_ns(1_000_000_000 + (i * 1_000_003) + 999)
    clamp = Instant.utc(2024, 1, 31) + months((i % 3) + 1)
    probe = seconds(5) + days(1) + anchor

    checksum += shifted.year * 37
    checksum += shifted.month * 31
    checksum += shifted.day * 29
    checksum += shifted.hour * 23
    checksum += shifted.minute * 19
    checksum += shifted.second * 17
    checksum += shifted.nanosecond % 1_000_003
    checksum += elapsed.total_nanoseconds % 1_000_003
    checksum += epoch.second + epoch.nanosecond % 1009
    checksum += tiny.unix_nanoseconds % 9973
    checksum += clamp.day * 13 + clamp.month
    checksum += probe.day + probe.second
    checksum += shifted.iso8601.length if (i % 97).zero?
    checksum += shifted > base ? 7 : 3
    checksum %= 2_147_483_647
  end

  checksum
end

puts main
